compiler-user-guide/src/developing-hardware/primitives.md

```diff
@@ -0,0 +1,397 @@
+      -> Signal dom a    -- din,  ARG[9]
+      -> Signal dom a
+  template: |-
+    -- blockRam begin
+    ~GENSYM[~RESULT_blockRam][1] : block
+      signal ~GENSYM[~RESULT_RAM][2] : ~TYP[5] := ~CONST[5];
+      signal ~GENSYM[rd][4]  : integer range 0 to ~LENGTH[~TYP[5]] - 1;
+      signal ~GENSYM[wr][5]  : integer range 0 to ~LENGTH[~TYP[5]] - 1;
+    begin
+      ~SYM[4] <= to_integer(~ARG[6])
+      -- pragma translate_off
+                    mod ~LENGTH[~TYP[5]]
+      -- pragma translate_on
+                    ;
+      ~SYM[5] <= to_integer(~ARG[8])
+      -- pragma translate_off
+                    mod ~LENGTH[~TYP[5]]
+      -- pragma translate_on
+                    ;
+    ~IF ~VIVADO ~THEN
+      ~SYM[6] : process(~ARG[3])
+      begin
+        if ~IF~ACTIVEEDGE[Rising][0]~THENrising_edge~ELSEfalling_edge~FI(~ARG[3]) then
+          if ~ARG[7] ~IF ~ISACTIVEENABLE[4] ~THEN and ~ARG[4] ~ELSE ~FI then
+            ~SYM[2](~SYM[5]) <= ~TOBV[~ARG[9]][~TYP[9]];
+          end if;
+          ~RESULT <= fromSLV(~SYM[2](~SYM[4]))
+          -- pragma translate_off
+          after 1 ps
+          -- pragma translate_on
+          ;
+        end if;
+      end process; ~ELSE
+      ~SYM[6] : process(~ARG[3])
+      begin
+        if ~IF~ACTIVEEDGE[Rising][0]~THENrising_edge~ELSEfalling_edge~FI(~ARG[3]) then
+          if ~ARG[7] ~IF ~ISACTIVEENABLE[4] ~THEN and ~ARG[4] ~ELSE ~FI then
+            ~SYM[2](~SYM[5]) <= ~ARG[9];
+          end if;
+          ~RESULT <= ~SYM[2](~SYM[4])
+          -- pragma translate_off
+          after 1 ps
+          -- pragma translate_on
+          ;
+        end if;
+      end process; ~FI
+    end block;
+    --end blockRam
+```
+
+Again, the `name` of the primitive is the fully qualified name of the function you are creating the primitive for.
+Because we are creating a _declaration_ primitive the _kind_ must be set to `Declaration`.
+Instead of discussing what the individual template holes mean in the above context, we will instead just give a general listing of the available template holes:
+
+* `~RESULT`: Signal to which the result of a primitive must be assigned
+  to. NB: Only used in a _declaration_ primitive.
+* `~ARG[N]`: `(N+1)`'th argument to the function.
+* `~CONST[N]`: `(N+1)`'th argument to the function. Like `~ARG`, but Clash will try to reduce this to a literal, even if it would otherwise consider it too expensive.
+  And if Clash fails to reduce this argument to a literal it will produce an error.
+* `~LIT[N]`: `(N+1)`'th argument to the function. Like `~CONST~` but values are   rendered as a bare literals, without any size or type annotations.
+  This only works for numeric types, and not for BitVector.
+* `~TYP[N]`: VHDL type of the `(N+1)`'th argument.
+* `~TYPO`: VHDL type of the result.
+* `~TYPM[N]`: VHDL type_name_ of the `(N+1)`'th argument; used in _type qualification_.
+* `~TYPMO`: VHDL type_name_ of the result; used in _type qualification_.
```

Comment:
```suggestion
* `~TYPM[N]`: VHDL type _name_ of the `(N+1)`'th argument; used in _type qualification_.
* `~TYPMO`: VHDL type _name_ of the result; used in _type qualification_.
```

There we go! When I wrote earlier that Markdown processors disagree on the semantics of an underscore in the middle of a word, I had no idea we were going to run into that problem a little further down already. mdBook doesn't render it as italics.

However, here it's a mistake that it's written as one word anyway.